### PR TITLE
added implementation of auth in server

### PR DESCRIPTION
**Pull request by sounak98**
- implemented login, signup and token generation
- implemented passport strategy for protected routes

### CODE REVIEW

**Comment by sounak98** [comment]
@Kwinten-C please review!

**Comment by Kwinten-C** [comment]
Thanks for submitting your bounty @sounak98 ! We will be looking into it as soon as possible.

**Comment by Kwinten-C** [comment]
Hi @sounak98 ! Looks good but there is no frontend and changePassword flow yet. But you can continue on this path. The rest looks fine.

**Comment by Kwinten-C** [comment]
Great! Any idea on when the front end flow will be completed?

**Comment by sounak98** [comment]
@Kwinten-C I've started working on the frontend, I'll update in a couple of days.

**Comment by sounak98** [comment]
@Kwinten-C I've updated the front-end!

The task-list currently stands as follows
- [x] User is able to register, login, change password using React UI component.
- [x] User authentication is verified using passport.js jwt strategy on the server-side and appropriate error or success message is displayed on UI
- [x] Corresponding entry is saved in mongodb using mongoose.js when user registers, logins, updates password etc.
- [ ] Test coverage for UI and Server-side test corresponding to this implementation is >70%

I'll update on the test coverage soon. Tests for the server have already been written, I just need to write tests for the UI.

File: server/test/test.js [hunk 1]
```diff
@@ -119,4 +128,99 @@ describe('running tests for API', function () {
       })
       .expect(500, /\"success\":false/ig, done);
   })
+
+  describe('auth', function() {
```
Comment (jernejpregelj) [inline server/test/test.js]: Move all tests out of "auth" describe and verify they run with "npm run test"

File: client/src/components/Profile.js [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+
+        return (
+            <div className="center">
+                <div className="card">
+                    <h1>Profile</h1>
+                    <form>
+                    {this.state.validationError && <div className="form-item-footer failure">{this.state.validationError}</div>}
+                        <input
+                            className="form-item"
+                            placeholder="New password goes here..."
```
Comment (jernejpregelj) [inline client/src/components/Profile.js]: Add entry for old password when changing passwords and check if it matches with previous one.

**Comment by Kwinten-C** [comment]
HI @sounak98 ! We reviewed it and asked for 2 change requests in the PR. Can you look into it?

**Comment by sounak98** [comment]
@Kwinten-C @jernejpregelj I have added the changes requested, please review again!

File: server/test/test.js (hunk 1 above)
Comment (sounak98) [reply server/test/test.js]: @jernejpregelj after proper setup, even the directly coupled tests pass.

screenshot 2019-02-02 at 12 37 50 am

**Comment by jernejpregelj** [comment]
@sounak98 Password change stuff is great and works! Still have problems with tests, please check my comment.

File: server/test/test.js (hunk 1 above)
Comment (sounak98) [reply server/test/test.js]: Tests are working absolutely fine with the command `npm run test`, on the same version (https://github.com/oceanprotocol/musicmap/commit/230b17a917a0e095079cde8df41f8cf7a9ae9547)
screenshot 2019-02-06 at 10 30 55 pm

**Pull request merged**